service/tablet_allocator.cc

```diff
@@ -149,6 +150,51 @@ struct migration_badness {
     bool operator<=>(const migration_badness& other) const = default;
 };
 
+struct colocated_tablets {
+    global_tablet_id left_tablet;
+    global_tablet_id right_tablet;
+
+    auto operator<=>(const colocated_tablets&) const = default;
+};
+
+// Represents either a single tablet replica or co-located replicas of sibling
+// tablets. The migration candidate is logically treated by balancer as a single
+// candidate. When candidate represents co-located replicas, it means that
+// the balancer will work to preserve the co-location by migrating those replicas
+// to same destination.
+struct migration_candidate {
```

Comment:
migration_tablet_set tablet_set could be good too. migration_tablet_set tablets makes it sound like there will always be multiple tablets in each candidate. 

See that each candidate might span one or more tablets, which is why I decided to name it migration_candidate, represents a logical candidate, since it might span two physical entities (tablets).

Otherwise, we'll have each node having candidates being typed as a set of migration_tablet_set (not very easy to understand why there's a set of a set in candidate list). And skipped candidate will also store a migration_tablet_set. skipped candidate becomes a logical term anyway, since it will store a tablet_set, which is in reality it can be two migration candidates (tablets).

Reply:
> migration_tablet_set tablet_set could be good too. migration_tablet_set tablets makes it sound like there will always be multiple tablets in each candidate.

plural doesn't imply that there must be more than one, it means there may be many. e.g. `_tables`, `_tablets`.  We can have 1 table and 1 tablet.

Reply:
> > migration_tablet_set tablet_set could be good too. migration_tablet_set tablets makes it sound like there will always be multiple tablets in each candidate.
> 
> plural doesn't imply that there must be more than one, it means there may be many. e.g. `_tables`, `_tablets`. We can have 1 table and 1 tablet.

This is a good point. It's very common in our field to do that naming when using sets.